app/uk/gov/hmrc/cgtpropertydisposals/controllers/actions/AuthenticatedAction.scala

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2019 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.cgtpropertydisposals.controllers.actions
+
+import java.time.LocalDateTime
+
+import com.google.inject.ImplementedBy
+import javax.inject.Inject
+import play.api.mvc._
+import uk.gov.hmrc.auth.core.retrieve._
+import uk.gov.hmrc.auth.core.{AuthConnector, AuthorisedFunctions, _}
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.HeaderCarrierConverter
+
+import scala.concurrent.{ExecutionContext, Future}
+
+final case class AuthenticatedUser(id: String)
+class AuthenticatedRequest[+A](
+  val user: AuthenticatedUser,
+  val timestamp: LocalDateTime,
+  val headerCarrier: HeaderCarrier,
+  request: Request[A]
+) extends WrappedRequest[A](request)
+
+@ImplementedBy(classOf[AuthenticateActionBuilder])
+trait AuthenticateActions extends ActionBuilder[AuthenticatedRequest, AnyContent]
+
+class AuthenticateActionBuilder @Inject()(
+  val authConnector: AuthConnector,
+  val parser: BodyParsers.Default,
+  val executionContext: ExecutionContext
+) extends AuthenticateActions
+    with AuthorisedFunctions {
+
+  private def extractHeaders(rh: RequestHeader): HeaderCarrier =
+    HeaderCarrierConverter.fromHeadersAndSessionAndRequest(rh.headers, Some(rh.session), Some(rh))
+
+  override def invokeBlock[A](request: Request[A], block: AuthenticatedRequest[A] => Future[Result]): Future[Result] = {
+    val forbidden = Results.Forbidden("Forbidden")
+    val carrier   = extractHeaders(request)
+    authorised()
+      .retrieve(v2.Retrievals.credentials) {
+        case Some(credentials) =>
```

Comment:
we should be making sure it's a GG login here so that we know it's a gg cred id we're dealing with. This can be done by chking that the provider type is `GovernmentGateway`. If we get `Verify` or `PrivilegedApplication` we want to log it and turn them away with a `Forbidden`